File: src/main/java/model/Hangman.java
package model;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class Hangman {

    private final List<Character> word;
    private final List<Character> triedChars;
    private final List<Character> wrongChars;
    private final List<Character> maskedWord;
    private final int WRONG_TRIES = 6;
    private Winner winner;

    public Hangman(String _word) {
        this.word = _word.toUpperCase().chars().mapToObj(e -> (char) e).collect(Collectors.toList());
        this.triedChars = new ArrayList<>();
        this.wrongChars = new ArrayList<>();
        this.maskedWord = new ArrayList<>();
        for (int i = 0; i < this.word.size(); i++) {
            this.maskedWord.add((this.word.get(i) == ' ') ? ' ' : (this.word.get(i) == '-') ? '-' : '_');
        }
    }

    public void check(char l) {
        char letter = Character.toUpperCase(l);
        if (letter < 'A' || letter > 'Z') {
            throw new IllegalArgumentException("Type a valid character.");
        }
        if (this.triedChars.stream().anyMatch(c -> c == letter)) {
            throw new IllegalArgumentException("This character was already pressed.");
        }
        this.triedChars.add(letter);
        this.maskedWord.clear();
        for (int i = 0; i < this.word.size(); i++) {
            char c = this.word.get(i);
            this.maskedWord.add(this.triedChars.contains(c) ? c : '_');
        }
        if (this.word.stream().anyMatch(c -> c == letter)) {
            if (this.maskedWord.stream().allMatch(c -> c != '_')) {
                this.winner = Winner.WIN;
                return;
            }
            this.winner = Winner.CORRECT_LETTER;
            return;
        }
        this.wrongChars.add(letter);
        if (this.wrongChars.size() == WRONG_TRIES) {
            this.winner = Winner.LOSE;
            return;
        }
        this.winner = Winner.WRONG_LETTER;
    }

    public List<Character> getWrongChars() {
        return this.wrongChars;
    }

    public List<Character> getWord() {
        return this.winner == Winner.LOSE ? this.word : this.maskedWord;
    }

    public Winner getWinner() {
        return winner;
    }
}

File: docs/Hangman.js
import Winner from "./Winner.js";

export default class Hangman {
    constructor(dictionary) {
        this.triedChars = [];
        this.wrongChars = [];
        this.maskedWord = [];
        this.WRONG_TRIES = 6;
        this.winner = null;
        this.word = this.selectWord(dictionary);
    }
    selectWord(words) {
        words = words.filter(e => e.length > 2);
        let x = Math.floor(Math.random() * words.length);
        let w = words[x].split('');
        this.maskedWord = w.map(c => (c === " ") ? " " : (c === "-") ? " &ndash; " : "_");
        return w;
    }
    getWord() {
        return this.winner == Winner.LOSE ? this.word : this.maskedWord;
    }
    getWrongChars() {
        return this.wrongChars;
    }
    getWinner() {
        return this.winner;
    }
    check(v) {
        if(this.winner === Winner.LOSE || this.winner === Winner.WIN) {
            return;
        }
        let letter = v.toUpperCase();
        if (letter < 'A' || letter > 'Z') {
            throw new Error("Type a valid character.");
        }
        if (this.triedChars.some(c => c === letter)) {
            throw new Error("This character was already pressed.");
        }
        this.triedChars.push(letter);
        this.maskedWord = [];
        for (const letter of this.word) {
            this.maskedWord.push(this.triedChars.includes(letter) ? letter : letter === '-' ? ' &ndash; ' : '_');
        }
        if (this.word.some(c => c === letter)) {
            this.winner = this.maskedWord.every(c => c !== "_") ? Winner.WIN : Winner.CORRECT_LETTER;
            return;
        }
        this.wrongChars.push(letter);
        this.winner = this.wrongChars.length == this.WRONG_TRIES ? Winner.LOSE : Winner.WRONG_LETTER;
    }
}